### DISCUSSION TITLE
What's the panic policy for dropping future and output?

### DISCUSSION THREAD

**Discussion by 3andero** [tokio-rs/tokio#4410] [General] [answered]
While it's clear that tokio will catch panics on polls, it has different policies for panics on drops.
For example:
If panic happens during the drop after completion, tokio will catch and ignore it:
https://github.com/tokio-rs/tokio/blob/4eed411519783ef6f58cbf74f886f91142b5cfa6/tokio/src/runtime/task/harness.rs#L292-L298

If panic happens during dropping the handle, tokio won't ignore it:
https://github.com/tokio-rs/tokio/blob/4eed411519783ef6f58cbf74f886f91142b5cfa6/tokio/src/runtime/task/harness.rs#L190-L192

If panic happens during cancellation, tokio will catch it:
https://github.com/tokio-rs/tokio/blob/4eed411519783ef6f58cbf74f886f91142b5cfa6/tokio/src/runtime/task/harness.rs#L431-L433

If panic happens after a panic during polling, tokio won't catch it and thus causes a double panic:
https://github.com/tokio-rs/tokio/blob/4eed411519783ef6f58cbf74f886f91142b5cfa6/tokio/src/runtime/task/harness.rs#L453-L459

Should we have a consistent policy for panics?

**Comment by Darksonn** [comment, marked as answer]
One of these is a bug, which I filed here: #4412. The rest are consistent behavior — the panic is passed on to the `JoinHandle` and the user can retrieve it by polling the `JoinHandle`. If the `JoinHandle` is dropped, then the panic is lost. (The panic handler will still print it to stdout.)

**Reply by 3andero** [reply]
Make sense except for the last case though. Don't we need to catch possible panic here?
https://github.com/tokio-rs/tokio/blob/4eed411519783ef6f58cbf74f886f91142b5cfa6/tokio/src/runtime/task/harness.rs#L453-L459

It seems the potential double panic will crash the whole runtime regardless of whether you're holding a `JoinHandle`.

**Reply by Darksonn** [reply]
Ah, well, I do think this one makes sense. If you panic, then all local variables are dropped, and the future behaves like a local variable in this case. A panic in the destructor of a local variable is a double panic.

**Reply by 3andero** [reply]
Now I understand this is a feature. Thank you for your explanation. 😊
